src/Npgsql/Internal/NpgsqlReadBuffer.cs

```diff
@@ -74,7 +74,87 @@ internal TimeSpan Timeout
         /// </summary>
         internal Encoding RelaxedTextEncoding { get; }
 
+#if DEBUG
+        internal bool EnableFrameTracking { get; set; }
+
+        internal void NextFrame()
+        {
+            Debug.Assert(_currentFrame.LengthLeft == 0, "Protocol failure");
+            _inFrame = false;
+        }
+
+        internal void ResetFrame()
+        {
+            _inFrame = true;
+            if (_previousFrame.HasValue)
+                _currentFrame = _previousFrame.Value;
+        }
+
+        bool _inFrame;
+
+        bool _bufferReset;
+
+        struct Frame
+        {
+            public Frame(byte code, int length, int lengthLeft)
+            {
+                Code = code;
+                Length = length;
+                LengthLeft = lengthLeft;
+            }
+
+            public readonly byte Code;
+            public readonly int Length;
+            public int LengthLeft;
+        }
+
+        Frame _currentFrame;
+        Frame? _previousFrame;
+
+        int _readPosition;
+
+        internal int ReadPosition
+        {
+            get => _readPosition;
+            set
+            {
+                if (EnableFrameTracking)
+                {
+                    var positionDelta = value - _readPosition;
+
+                    // Setting the same position as the one we already have
+                    if (positionDelta == 0)
+                        return;
+
+                    // Buffer reset
+                    if (!_bufferReset)
+                    {
+                        if (!_inFrame)
+                        {
+                            Debug.Assert(positionDelta == 1, "Protocol failure");
+                            Debug.Assert(ReadBytesLeft >= 5);
+                            var frameCode = Buffer[_readPosition];
+                            var frameLength = BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<int>(ref Buffer[_readPosition + 1])) + 1;
+                            var frameLengthLeft = frameLength - 1;
+                            _previousFrame = _currentFrame;
+                            _currentFrame = new Frame(frameCode, frameLength, frameLengthLeft);
+                            _inFrame = true;
+                        }
+                        else
+                        {
+                            _currentFrame.LengthLeft -= positionDelta;
```

Comment:
nit: maybe better to have Position (in the frame) - just like the the buffer's ReadPosition - which gets incremented, rather than LengthLeft which gets decremented.